atomicfu/src/jvmMain/kotlin/kotlinx/atomicfu/Trace.kt

```diff
@@ -0,0 +1,26 @@
+@file:Suppress("INVISIBLE_REFERENCE", "INVISIBLE_MEMBER")
+
+package kotlinx.atomicfu
+
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.internal.InlineOnly
+
+@Suppress("FunctionName")
+@InlineOnly
+public actual fun Trace(size: Int, format: (Int, String) -> String): TraceBase =
+    TraceImpl(size, format)
+
+private class TraceImpl(size: Int, val format: (Int, String) -> String) : TraceBase() {
+    init { require(size >= 1) }
+    private val size = ((size shl 1) - 1).takeHighestOneBit() // next power of 2
+    private val mask = this.size - 1
+    private val trace = arrayOfNulls<String>(this.size)
+    private val index = AtomicInteger(0)
+
+    override fun append(text: String) {
+        val i = index.getAndIncrement()
+        trace[i and mask] = format(index.get(), text)
+    }
+
+    override fun toString(): String = trace.filterNotNull().joinToString()
```

Comment:
This is Ok for tests, but no so good for real use-code that might want to dump trace to the console in case when something went wrong. It would be hard to parse when trace index had rolled over, so I'd suggest the following improvements:
* Start `toString` from the current `index` in array (then wrap around the end), so that all messages get output in their chronological order. 
* Use `/n` (LF) as a separator. This way it is going to look nice if it gets printed to any kind of log.